include/vara/Feature/ConstraintBuilder.h

```diff
@@ -0,0 +1,342 @@
+    }
+
+  private:
+    ValueTy Value;
+  };
+
+  template <class ConstraintTy>
+  class UnaryBuilderImpl : public ConstraintBuilderImpl {
+
+  public:
+    UnaryBuilderImpl() : Operand(nullptr) {}
+
+    std::unique_ptr<ConstraintBuilderImpl> *next() override { return &Operand; }
+
+    std::unique_ptr<Constraint> build() override {
+      return std::make_unique<ConstraintTy>(std::move(Operand->build()));
+    }
+
+  private:
+    std::unique_ptr<ConstraintBuilderImpl> Operand;
+  };
+
+  template <class ConstraintTy>
+  class BinaryBuilderImpl : public ConstraintBuilderImpl {
+
+  public:
+    BinaryBuilderImpl(std::unique_ptr<ConstraintBuilderImpl> LeftOperand)
+        : LeftOperand(std::move(LeftOperand)), RightOperand(nullptr) {}
+
+    std::unique_ptr<ConstraintBuilderImpl> *next() override {
+      return &RightOperand;
+    }
+
+    std::unique_ptr<Constraint> build() override {
+      return std::make_unique<ConstraintTy>(std::move(LeftOperand->build()),
+                                            std::move(RightOperand->build()));
+    }
+
+  private:
+    std::unique_ptr<ConstraintBuilderImpl> LeftOperand;
+    std::unique_ptr<ConstraintBuilderImpl> RightOperand;
+  };
+
+  template <class ConstraintTy>
+  ConstraintBuilder &unary() {
+    if (!Head || *Head) {
+      llvm::errs() << "Syntax error: Unrecognized unary constraint.\n";
+      Head = nullptr;
+      return *this;
+    }
+    *Head = std::make_unique<UnaryBuilderImpl<ConstraintTy>>();
+    Head = (*Head)->next();
+    return *this;
+  }
+
+  template <class ConstraintTy>
+  ConstraintBuilder &binary() {
+    if (!Head || !*Head) {
+      llvm::errs() << "Syntax error: Unrecognized binary constraint.\n";
+      Head = nullptr;
+      return *this;
+    }
+    *Head = std::make_unique<BinaryBuilderImpl<ConstraintTy>>(std::move(*Head));
+    Head = (*Head)->next();
+    return *this;
+  }
+
+public:
+  /// Builds constraint.
+  ///
+  /// \return unique pointer for built constraint
+  [[nodiscard]] std::unique_ptr<Constraint> build() {
+    if (!Head || !*Head || !Root || !Parentheses.empty()) {
+      llvm::errs() << "Syntax error: Incomplete constraint.\n";
+      return nullptr;
+    }
+    return Root->build();
+  }
+
+  /// Parenthesizes intermediate constraint.
+  ///
+  /// Example:
+  ///     feature("A").implies().feature("B")()
+  ///      .implies().feature("C").implies().feature("D")
+  ///   produces the constraint
+  ///     '((A => B) => (C => D))'
+  ConstraintBuilder &operator()() {
+    if (!Head || !*Head || !Parentheses.empty()) {
+      llvm::errs() << "Syntax error: Unrecognized parentheses.\n";
+      Head = nullptr;
+      return *this;
+    }
+    Head = &Root;
+    return *this;
+  }
+
+  /// Open parenthesis. Allows to add more precise parentheses.
+  ///
+  /// Example:
+  ///     left().feature("A").implies().feature("B").right()
+  ///      .implies().feature("C").implies().feature("D")
```

Comment:
Why does this example have not left/right/ for the right side `(C => D)`

Reply:
Using the builder all operators are parsed as right-associative by default and parentheses inserted respectively (we currently do not have precedence to keep it simple), so we do not need to specify those here. It would be fine if one would do so, but not necessary.

Reply:
How would you, with this syntax then express this expression: `((A => B) => C => D)` and  `(((A => B) => C) => D)`

Reply:
You would need too explicitly parenthesize both expressions or simply use the `()` operator.

`feature("A").implies().feature("B")().implies().feature("C")().implies().feature("D")`